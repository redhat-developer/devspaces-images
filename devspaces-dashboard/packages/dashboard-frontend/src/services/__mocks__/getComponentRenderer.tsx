/*
 * Copyright (c) 2018-2024 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import { render } from '@testing-library/react';
import React from 'react';
import renderer from 'react-test-renderer';

export * from '@testing-library/react';

export default function <T extends Array<unknown>>(
  getComponent: (..._args: T) => React.ReactElement,
): {
  createSnapshot: (...args: T) => renderer.ReactTestRenderer;
  renderComponent: (...args: T) => {
    reRenderComponent: (..._args: T) => void;
  };
} {
  return {
    createSnapshot: (...args: T) => {
      return renderer.create(getComponent(...args));
    },
    renderComponent: (...args: T) => {
      const res = render(getComponent(...args));
      return {
        reRenderComponent: (..._args: T) => {
          res.rerender(getComponent(..._args));
        },
      };
    },
  };
}
