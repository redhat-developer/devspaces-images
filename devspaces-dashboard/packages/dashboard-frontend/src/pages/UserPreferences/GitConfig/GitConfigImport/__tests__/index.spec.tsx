/*
 * Copyright (c) 2018-2024 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import { ValidatedOptions } from '@patternfly/react-core';
import userEvent from '@testing-library/user-event';
import React from 'react';

import {
  fieldId,
  fileNamePlaceholder,
  GitConfigImport,
  textAreaPlaceholder,
} from '@/pages/UserPreferences/GitConfig/GitConfigImport';
import getComponentRenderer, { screen, waitFor } from '@/services/__mocks__/getComponentRenderer';

const { renderComponent, createSnapshot } = getComponentRenderer(getComponent);

const mockOnChange = jest.fn();

describe('GitConfigUpload', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  test('snapshot without content, validation is in the default state', () => {
    const snapshot = createSnapshot(ValidatedOptions.default);
    expect(snapshot.toJSON()).toMatchSnapshot();
  });

  test('snapshot with content, validation is in the default state', () => {
    const snapshot = createSnapshot(
      ValidatedOptions.default,
      '[user]\n\tname = User One\n\temail = user-1@chetest.com\n',
    );
    expect(snapshot.toJSON()).toMatchSnapshot();
  });

  test('snapshot, validation is in the error state', () => {
    const snapshot = createSnapshot(ValidatedOptions.error);
    expect(snapshot.toJSON()).toMatchSnapshot();
  });

  test('File uploader controls', () => {
    renderComponent(ValidatedOptions.default);

    const fileUploader = screen.queryByTestId(fieldId);

    // file uploader is rendered
    expect(fileUploader).not.toBeNull();

    // file input field
    const fileInput = screen.queryByPlaceholderText(fileNamePlaceholder);
    expect(fileInput).not.toBeNull();
    expect(fileInput).toHaveAttribute('readonly');

    // Upload button
    const uploadButton = screen.queryByText('Upload');
    expect(uploadButton).not.toBeNull();
    expect(uploadButton).toBeEnabled();

    // Clear button
    const clearButton = screen.queryByText('Clear');
    expect(clearButton).not.toBeNull();
    expect(clearButton).toBeDisabled();

    // content input field
    const contentInput = screen.queryByPlaceholderText(textAreaPlaceholder);
    expect(contentInput).not.toBeNull();
    expect(contentInput).not.toHaveAttribute('readonly');
    expect(contentInput).toBeEnabled();
  });

  describe('Upload file', () => {
    test('should handle a valid file', async () => {
      renderComponent(ValidatedOptions.default);

      const fileInput = screen.getByPlaceholderText(fileNamePlaceholder);

      const clearButton = screen.getByText('Clear');

      /* Upload a file */

      const fileContent = '[user]\n\tname = User One\n\temail = user-1@chetest.com\n';
      const file = new File([fileContent], '.gitconfig', { type: 'text/plain' });

      const fileUploader = screen.getByTestId(fieldId);
      const fileUploadInput = fileUploader.querySelector('input[type="file"]');
      userEvent.upload(fileUploadInput!, file);

      await waitFor(() => expect(mockOnChange).toHaveBeenCalledWith(fileContent, true));

      expect(fileInput).toHaveValue(file.name);
      expect(clearButton).toBeEnabled();

      /* Clear the field */

      userEvent.click(clearButton);

      expect(fileInput).toHaveValue('');
      expect(clearButton).toBeDisabled();
    });
  });

  describe('Paste content', () => {
    test('should handle a valid content', async () => {
      renderComponent(ValidatedOptions.default);

      const contentInput = screen.getByPlaceholderText(textAreaPlaceholder);

      const clearButton = screen.getByText('Clear');

      const fileInput = screen.getByPlaceholderText(fileNamePlaceholder);

      /* Paste a content */

      const content = 'content';
      userEvent.paste(contentInput, content);

      await waitFor(() => expect(mockOnChange).toHaveBeenCalledWith(content, false));

      expect(fileInput).toHaveValue('');
      expect(clearButton).toBeEnabled();
      expect(contentInput).toHaveTextContent(content);
      expect(contentInput).not.toHaveAttribute('readonly');

      /* Clear the field */

      userEvent.click(clearButton);

      expect(contentInput).toHaveValue('');
      expect(clearButton).toBeDisabled();
    });
  });
});

function getComponent(validated: ValidatedOptions, content?: string): React.ReactElement {
  return <GitConfigImport content={content} validated={validated} onChange={mockOnChange} />;
}
