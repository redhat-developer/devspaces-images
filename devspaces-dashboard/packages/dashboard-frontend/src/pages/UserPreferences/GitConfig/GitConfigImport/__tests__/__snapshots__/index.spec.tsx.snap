// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GitConfigUpload snapshot with content, validation is in the default state 1`] = `
<div
  className="pf-c-file-upload"
  data-testid="git-configuration"
  onBlur={[Function]}
  onClick={[Function]}
  onDragEnter={[Function]}
  onDragLeave={[Function]}
  onDragOver={[Function]}
  onDragStart={[Function]}
  onDrop={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  tabIndex={null}
>
  <div
    className="pf-c-file-upload__file-select"
  >
    <div
      className="pf-c-input-group"
    >
      <input
        aria-describedby="git-configuration-browse-button"
        aria-invalid={false}
        aria-label="Upload the .gitconfig file"
        className="pf-c-form-control"
        data-ouia-component-id="OUIA-Generated-TextInputBase-2"
        data-ouia-component-type="PF4/TextInput"
        data-ouia-safe={true}
        disabled={false}
        id="git-configuration-filename"
        name="git-configuration-filename"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        placeholder="Upload the .gitconfig file"
        readOnly={true}
        required={false}
        type="text"
        value=""
      />
      <button
        aria-disabled={false}
        aria-label={null}
        className="pf-c-button pf-m-control"
        data-ouia-component-id="OUIA-Generated-Button-control-3"
        data-ouia-component-type="PF4/Button"
        data-ouia-safe={true}
        disabled={false}
        id="git-configuration-browse-button"
        onClick={[Function]}
        role={null}
        type="button"
      >
        Upload
      </button>
      <button
        aria-disabled={false}
        aria-label={null}
        className="pf-c-button pf-m-control"
        data-ouia-component-id="OUIA-Generated-Button-control-4"
        data-ouia-component-type="PF4/Button"
        data-ouia-safe={true}
        disabled={false}
        onClick={[Function]}
        role={null}
        type="button"
      >
        Clear
      </button>
    </div>
  </div>
  <div
    className="pf-c-file-upload__file-details"
  >
    <textarea
      aria-invalid={false}
      aria-label="File upload"
      className="pf-c-form-control pf-m-resize-vertical"
      disabled={false}
      id="git-configuration"
      name="git-configuration"
      onChange={[Function]}
      onClick={[Function]}
      placeholder="Or paste the Git Configuration here"
      readOnly={false}
      required={true}
      value="[user]
	name = User One
	email = user-1@chetest.com
"
    />
  </div>
  <input
    autoComplete="off"
    multiple={false}
    onChange={[Function]}
    onClick={[Function]}
    style={
      {
        "display": "none",
      }
    }
    tabIndex={-1}
    type="file"
  />
</div>
`;

exports[`GitConfigUpload snapshot without content, validation is in the default state 1`] = `
<div
  className="pf-c-file-upload"
  data-testid="git-configuration"
  onBlur={[Function]}
  onClick={[Function]}
  onDragEnter={[Function]}
  onDragLeave={[Function]}
  onDragOver={[Function]}
  onDragStart={[Function]}
  onDrop={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  tabIndex={null}
>
  <div
    className="pf-c-file-upload__file-select"
  >
    <div
      className="pf-c-input-group"
    >
      <input
        aria-describedby="git-configuration-browse-button"
        aria-invalid={false}
        aria-label="Upload the .gitconfig file"
        className="pf-c-form-control"
        data-ouia-component-id="OUIA-Generated-TextInputBase-1"
        data-ouia-component-type="PF4/TextInput"
        data-ouia-safe={true}
        disabled={false}
        id="git-configuration-filename"
        name="git-configuration-filename"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        placeholder="Upload the .gitconfig file"
        readOnly={true}
        required={false}
        type="text"
        value=""
      />
      <button
        aria-disabled={false}
        aria-label={null}
        className="pf-c-button pf-m-control"
        data-ouia-component-id="OUIA-Generated-Button-control-1"
        data-ouia-component-type="PF4/Button"
        data-ouia-safe={true}
        disabled={false}
        id="git-configuration-browse-button"
        onClick={[Function]}
        role={null}
        type="button"
      >
        Upload
      </button>
      <button
        aria-disabled={true}
        aria-label={null}
        className="pf-c-button pf-m-control pf-m-disabled"
        data-ouia-component-id="OUIA-Generated-Button-control-2"
        data-ouia-component-type="PF4/Button"
        data-ouia-safe={true}
        disabled={true}
        onClick={[Function]}
        role={null}
        tabIndex={null}
        type="button"
      >
        Clear
      </button>
    </div>
  </div>
  <div
    className="pf-c-file-upload__file-details"
  >
    <textarea
      aria-invalid={false}
      aria-label="File upload"
      className="pf-c-form-control pf-m-resize-vertical"
      disabled={false}
      id="git-configuration"
      name="git-configuration"
      onChange={[Function]}
      onClick={[Function]}
      placeholder="Or paste the Git Configuration here"
      readOnly={false}
      required={true}
      value=""
    />
  </div>
  <input
    autoComplete="off"
    multiple={false}
    onChange={[Function]}
    onClick={[Function]}
    style={
      {
        "display": "none",
      }
    }
    tabIndex={-1}
    type="file"
  />
</div>
`;

exports[`GitConfigUpload snapshot, validation is in the error state 1`] = `
<div
  className="pf-c-file-upload"
  data-testid="git-configuration"
  onBlur={[Function]}
  onClick={[Function]}
  onDragEnter={[Function]}
  onDragLeave={[Function]}
  onDragOver={[Function]}
  onDragStart={[Function]}
  onDrop={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  tabIndex={null}
>
  <div
    className="pf-c-file-upload__file-select"
  >
    <div
      className="pf-c-input-group"
    >
      <input
        aria-describedby="git-configuration-browse-button"
        aria-invalid={false}
        aria-label="Upload the .gitconfig file"
        className="pf-c-form-control"
        data-ouia-component-id="OUIA-Generated-TextInputBase-3"
        data-ouia-component-type="PF4/TextInput"
        data-ouia-safe={true}
        disabled={false}
        id="git-configuration-filename"
        name="git-configuration-filename"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        placeholder="Upload the .gitconfig file"
        readOnly={true}
        required={false}
        type="text"
        value=""
      />
      <button
        aria-disabled={false}
        aria-label={null}
        className="pf-c-button pf-m-control"
        data-ouia-component-id="OUIA-Generated-Button-control-5"
        data-ouia-component-type="PF4/Button"
        data-ouia-safe={true}
        disabled={false}
        id="git-configuration-browse-button"
        onClick={[Function]}
        role={null}
        type="button"
      >
        Upload
      </button>
      <button
        aria-disabled={true}
        aria-label={null}
        className="pf-c-button pf-m-control pf-m-disabled"
        data-ouia-component-id="OUIA-Generated-Button-control-6"
        data-ouia-component-type="PF4/Button"
        data-ouia-safe={true}
        disabled={true}
        onClick={[Function]}
        role={null}
        tabIndex={null}
        type="button"
      >
        Clear
      </button>
    </div>
  </div>
  <div
    className="pf-c-file-upload__file-details"
  >
    <textarea
      aria-invalid={true}
      aria-label="File upload"
      className="pf-c-form-control pf-m-resize-vertical"
      disabled={false}
      id="git-configuration"
      name="git-configuration"
      onChange={[Function]}
      onClick={[Function]}
      placeholder="Or paste the Git Configuration here"
      readOnly={false}
      required={true}
      value=""
    />
  </div>
  <input
    autoComplete="off"
    multiple={false}
    onChange={[Function]}
    onClick={[Function]}
    style={
      {
        "display": "none",
      }
    }
    tabIndex={-1}
    type="file"
  />
</div>
`;
